components/core/src/os/signals/unix.rs

```diff
@@ -1,87 +1,149 @@
 //! Traps and notifies UNIX signals.
 
-use crate::os::process::{Signal,
-                         SignalCode};
-use std::{collections::VecDeque,
-          sync::{atomic::Ordering,
-                 Mutex,
-                 Once}};
+use crate::os::process::SignalCode;
+use std::{io,
+          mem,
+          ptr,
+          sync::{atomic::{AtomicBool,
+                          Ordering},
+                 Once},
+          thread};
 
 static INIT: Once = Once::new();
+static PENDING_HUP: AtomicBool = AtomicBool::new(false);
+static PENDING_CHLD: AtomicBool = AtomicBool::new(false);
 
-lazy_static::lazy_static! {
-    static ref CAUGHT_SIGNALS: Mutex<VecDeque<SignalCode>> = Mutex::new(VecDeque::new());
+pub fn init() {
+    INIT.call_once(|| {
+            // TODO(ssd) 2019-10-16: We could bubble this error up
+            // further if we want, but in either case this should be a
+            // hard failure.
+            self::start_signal_handler().expect("starting signal handler failed");
+        });
 }
 
-// Functions from POSIX libc.
-extern "C" {
-    fn signal(sig: SignalCode,
-              cb: unsafe extern "C" fn(SignalCode))
-              -> unsafe extern "C" fn(SignalCode);
-}
+pub fn pending_sighup() -> bool { PENDING_HUP.compare_and_swap(true, false, Ordering::SeqCst) }
+pub fn pending_sigchld() -> bool { PENDING_CHLD.compare_and_swap(true, false, Ordering::SeqCst) }
```

Comment:
Reviewer Note:  I've gone with `SeqCst` here because that is what we did for SHUTDOWN and it leaves no room for doubt.  But I think we may be able to use Relaxed here.